### PR TITLE
206 follow system

### LINKED ISSUE #206: [206] Following system
**Current state**
There is no form of community that enables users to follow each other

**Expected state**
Users are able to follow each other. When you follow a user it says 'Following' on their profile. When they follow you too it says 'Friends'.

**TO DO**
- [ ] Follow user (Back-end)
- [ ] Follow user (Front-end)
- [ ] Follower count on profile

**Necessities**
None

**Branch**
206-follow-system

### PR DESCRIPTION
**Pull request by AjobK**
closes #206 

**What is it supposed to do**
Users should be able to follow other users to see activity in the future

**How can we test your branch**
1. Go to a profile
2. Click the follow button
3. Check if the follow is still active when moving around in the app
4. Unfollow
5. Check if the unfollow is still active when moving around in the app
6. Check responsiveness
7. Etc.

**What broweser have been checked**
- [x] Chrome
- [x] Firefox
- [x] Opera
- [x] Edge
- [ ] Safari

### CODE REVIEW

File: frontend/src/components/userBanner/index.js [hunk 1]
```diff
@@ -23,9 +45,15 @@ class UserBanner extends Component {
         <div className={styles.innerWrapper}>
```
Comment (MrGooss) [inline frontend/src/components/userBanner/index.js]: spaces between curly braces (entire file)

File: frontend/src/components/userBanner/index.js [hunk 2]
```diff
@@ -2,9 +2,31 @@ import React, { Component } from 'react'
...
+
+    this.state = {
+      following: this.props.user.following || false
+    }
+  }
+
+  follow = () => {
+    const username = window.location.pathname.split('/').filter(i => i != '').pop()
+
+    Axios.post(`${this.props.store.defaultData.backendUrl}/profile/follow/${username}`, {}, {withCredentials: true})
```
Comment (MrGooss) [inline frontend/src/components/userBanner/index.js]: Spaces between curly braces

**Review by MrGooss — CHANGES_REQUESTED** [review]

File: frontend/src/components/userBanner/index.js [hunk 3]
```diff
@@ -2,9 +2,31 @@ import React, { Component } from 'react'
...
+    .then((res) => {
+      console.log(res.data.following)
```
Comment (ryankroon00) [inline frontend/src/components/userBanner/index.js]: Remove console.log

File: backend/app/dao/profileDAO.ts [hunk 4]
```diff
@@ -1,5 +1,7 @@
-import DatabaseConnector from '../util/databaseConnector';
-import { Profile } from '../entity/profile';
+import DatabaseConnector from '../util/databaseConnector'
+import { Profile } from '../entity/profile'
+import ProfileFollowedBy from '../entity/profile_followed_by'
+import { profile } from 'console'
```
Comment (ryankroon00) [inline backend/app/dao/profileDAO.ts]: Declared but never used

File: backend/app/service/profileService.ts [hunk 5]
```diff
@@ -141,6 +153,46 @@ class ProfileService {
...
+        if (req.params.username)
+            profile = await this.dao.getProfileByUsername(req.params.username)
+
+        if (!follower || !profile)
+            return res.status(422).json({ error: 'No profile to follow or follower found in request' })
+
+        if (follower.id == profile.id)
+            return res.status(422).json({ error: 'Following yourself is not possible' })
+
+        let profileFollowedBy: ProfileFollowedBy = new ProfileFollowedBy()
```
Comment (ryankroon00) [inline backend/app/service/profileService.ts]: Is never reassigned so use const instead of let

File: backend/app/service/profileService.ts [hunk 6]
```diff
@@ -74,8 +75,19 @@ class ProfileService {
         if (receivedUsername && decodedToken)
             isOwner = !!(receivedUsername == decodedToken.username)
 
+        let following = false
+
+        if (!isOwner && decodedToken && decodedToken.username) {
+            let profileFollowedBy: ProfileFollowedBy = new ProfileFollowedBy()
```
Comment (ryankroon00) [inline backend/app/service/profileService.ts]: Never reassigned please use const instead of let

File: backend/app/service/profileService.ts [hunk 7]
```diff
@@ -141,6 +153,46 @@ class ProfileService {
...
+        let follower: Profile | null = null
+
+        if (decodedToken && decodedToken.username) {
+            follower = await this.dao.getProfileByUsername(decodedToken.username)
+        } else {
+            return res.status(404).json({ error: 'No username to follow was given' })
+        }
+
+        let profile: Profile | null = null
+
```
Comment (ryankroon00) [inline backend/app/service/profileService.ts]: Trailing spaces not allowed

File: backend/app/service/profileService.ts [hunk 8]
```diff
@@ -141,6 +153,46 @@ class ProfileService {
...
+        if (!follower || !profile)
+            return res.status(422).json({ error: 'No profile to follow or follower found in request' })
+
+        if (follower.id == profile.id)
+            return res.status(422).json({ error: 'Following yourself is not possible' })
+
+        let profileFollowedBy: ProfileFollowedBy = new ProfileFollowedBy()
+        profileFollowedBy.follower = follower.id
+        profileFollowedBy.profile = profile.id
+
```
Comment (ryankroon00) [inline backend/app/service/profileService.ts]: Trailing spaces not allowed

**Review by ryankroon00 — CHANGES_REQUESTED** [review]
Working great, just a few small fixes and it can be merged <3

File: backend/app/controllers/authorizationController.ts [hunk 9]
```diff
@@ -1,23 +1,134 @@
...
+            res.status(401).json({
+                error: 'Invalid jwt'
+            })
+        }
     }
 }
-export default AuthorizationController
\ No newline at end of file
+
+export default AuthorizationController
```
Comment (MrGooss) [inline backend/app/controllers/authorizationController.ts]: New line at end of file

File: backend/app/controllers/postController.ts [hunk 10]
```diff
@@ -218,4 +218,4 @@ class PostService {
     }
 }
 
-export default PostService
+export default PostController
```
Comment (MrGooss) [inline backend/app/controllers/postController.ts]: Newline at end of file

File: backend/app/controllers/profileController.ts [hunk 11]
```diff
@@ -271,4 +271,4 @@ class ProfileService {
         return account
     }
 }
-export default ProfileService
+export default ProfileController
```
Comment (MrGooss) [inline backend/app/controllers/profileController.ts:275]: Newline at end of file

File: backend/app/dao/commentDAO.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+            })
+
+            repository.save(children)
+        }
+
+        return repository.save(comment)
+    }
+}
+
+export default CommentDAO
```
Comment (MrGooss) [inline backend/app/dao/commentDAO.ts]: Newline at end of file

File: backend/app/entity/comment.ts [hunk 13]
```diff
@@ -17,4 +30,4 @@ export class Comment extends BaseEntity {
     @Column({ nullable: true })
     archived_at: Date
 }
-export default Comment
\ No newline at end of file
+export default Comment
```
Comment (MrGooss) [inline backend/app/entity/comment.ts:34]: Newline at end of file

File: backend/app/routes/authorizationRoutes.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+        this.initRoutes()
+    }
+
+    public initRoutes(): void {
+        this.router.get(this.loginVerify, this.authController.loginVerify)
+        this.router.post(this.login, this.authController.login)
+        this.router.get(this.logout, this.authController.logout)
+    }
+}
+export default AuthorizationRoutes
```
Comment (MrGooss) [inline backend/app/routes/authorizationRoutes.ts]: Newline at end of file

File: backend/app/routes/authorizationRoutes.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+        this.authController = new AuthorizationController()
+        this.initRoutes()
+    }
+
+    public initRoutes(): void {
+        this.router.get(this.loginVerify, this.authController.loginVerify)
+        this.router.post(this.login, this.authController.login)
+        this.router.get(this.logout, this.authController.logout)
+    }
+}
```
Comment (MrGooss) [inline backend/app/routes/authorizationRoutes.ts:22]: Newline

File: backend/app/routes/commentRoutes.ts [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    public initRoutes(): void {
+        this.router.get(this.comment + '/:path', this.commentController.getComments)
+        this.router.post(this.comment, isAuth, this.commentController.createComment)
+        this.router.delete(this.comment + '/:id', isAuth, this.commentController.deleteComment)
+    }
+}
+
+export default CommentRoutes
```
Comment (MrGooss) [inline backend/app/routes/commentRoutes.ts]: Newline at end of file

File: backend/app/routes/postRoutes.ts [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+        this.router.get(this.post + '/owned-by/:username', this.postController.getOwnedPosts)
+
+        this.router.post(this.post + '/like/:path', auth, this.postController.likePost)
+        this.router.delete(this.post + '/like/:path', auth, this.postController.unlikePost)
+        this.router.get(this.post + '/like/:path', this.postController.getPostLikes)
+        this.router.get(this.post + '/liked-by/recent/:username', this.postController.getRecentUserLikes)
+        this.router.put(this.post + '/:path', auth, this.postController.updatePost)
+    }
+}
```
Comment (MrGooss) [inline backend/app/routes/postRoutes.ts:30]: Newline

File: backend/app/routes/postRoutes.ts [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+export default PostRoutes
```
Comment (MrGooss) [inline backend/app/routes/postRoutes.ts:32]: Newline at end of file

File: backend/app/routes/profileRoutes.ts [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+        this.initRoutes()
+    }
+
+    public initRoutes(): void {
+        this.router.post(this.register, this.profileController.register)
+        this.router.put(this.profile, auth, this.profileController.updateProfile)
+        this.router.get (this.profile, this.profileController.getProfile)
+        this.router.post(this.follow, this.profileController.follow)
+    }
+}
```
Comment (MrGooss) [inline backend/app/routes/profileRoutes.ts:25]: Newline

File: backend/app/routes/profileRoutes.ts [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+export default ProfileRoutes
```
Comment (MrGooss) [inline backend/app/routes/profileRoutes.ts]: Newline at end of file

**Review by MrGooss — CHANGES_REQUESTED** [review]

**Pull request merged**